lib/string_basics.dart

```diff
@@ -200,6 +200,32 @@ extension StringBasics on String {
     }
     return stringBuffer.toString();
   }
+
+  /// Returns a truncated version of the string.
+  ///
+  /// Example:
+  /// ```dart
+  /// final sentence = 'The quick brown fox jumps over the lazy dog';
+  /// final truncated = sentence.limit(20); // 'The quick brown fox...'
+  /// ```
+  ///
+  /// The [limit] is the truncated length of the string.
+  /// The [end] is the suffix or ending string of the truncated string.
+  /// The [trim] is whether or not to trim the spaces of the truncated string
+  /// before appending the ending string.
+  String limit(int limit, {String end = '...', bool trim = true}) {
+    if (this.length <= limit) {
+      return this;
+    }
+
+    String truncated = this.slice(start: 0, end: limit);
```

Comment:
This doesn't handle non-ASCII characters properly: `'👍👍👍👍👍'.limit(3)` produces `'👍�...'` instead of a desired `'👍👍…'` or `'👍👍👍…'` (see [the other conversation](https://github.com/google/dart-basics/pull/52/files#r917444585)). This can be corrected by using the [<kbd>characters</kbd>](https://pub.dev/packages/characters) package:

```dart// Shortcut to avoid having to traverse individual human-perceived characters
// (grapheme clusters) of the string. This works because each character uses one
// or more code units, so `this.length` is an upper bound to the number of
// characters.
if (this.length <= limit) return this;

// Calculating the number of characters of a string is an operation with linear
// time. For a string with a length of multiple kilobytes and a `limit` of
// 10, this method only needs to traverse 11 characters instead of the whole
// string.
//
// We have to take one more character than `limit`: If the resulting
// `substring.length` is `limit` or less, the string was shorter than the limit.
// Otherwise, `substring.length` is equal to `limit + 1` and the original string
// is longer, meaning we have to add the ellipsis (`end`).
//
// (`substring.length` calls `Characters.length`, i.e., gets the length of the
// `Characters` iterable, so it counts human-perceived characters and not code
// units.)
final substring = this.characters.take(limit + 1);
if (substring.length <= limit) return this;

// Finally, we can remove the additional character, make space for the ellipsis
// (if desired, see a different conversation), and append the ellipsis (`end`).
final endLength = end.characters.length;
return '${substring.skipLast(endLength + 1)}$end';
```

This snippet produces a string whose total length, including the ellipsis, is at most `limit`. It might have to be adjusted depending on the outcome of the other conversation linked above.

Reply:
Hi @zdg2102 

Is it advisable or allowed to use third party packages? I don't mind updating the code and follow the suggestion in this comment but I am not sure if it is ok to use any third party package.

Reply:
We used to use `quiver` (we ended up dropping the dependency to make the non-null migration easier), so I don't think there's any issue with using third party code in general. I personally think this sounds like a reasonable use case, since this is definitely logic we wouldn't want to be re-implementing in `basics`.

@johnsonmh any objections to adding a dependency on `characters` here?

Reply:
Hi @zdg2102 

I tried the [characters package](https://pub.dev/packages/characters) and it does work and will definitely resolve the issue so I used it.